### ISSUE TITLE
Changing Overlay Axis label based on Dimension's  unit attribute

### ISSUE THREAD

**Issue by johnzzzz** [ioam/holoviews#2162]
When composing multiple Curves into a single chart using an Overlay, the y-axis is formatted as the FIRST Curve's  ydims Dimension's  "{label} ({unit})". 
This labeling convention might be fine to display a single Curve, but if there are multiple Curves per chart, the Y-axis label should not mention the First curve's label, but should instead be the Units that are common to all the Curves in the Overlay. 
1. How does one directly change the axis label of an Overlay?
2. Is there a way to get the Dimensions units to be used as an axis label?
3. Is there control over the axis label's format?

For example this code creates a chart comparing the acceleration  of two cars. The y-axis should be the unit "mph" and not "Porsche (mph)"

``` python
# matplotlib defaults
import matplotlib as mpl
mpl.rcParams['figure.facecolor']="white"
# Holoviews
import holoviews as hv
hv.extension('matplotlib')
# Compare Accelleration Data
import pandas as pd
raw_data = [{'time':0, 'yugo':0, 'porsche':0},
            {'time':1, 'yugo':10, 'porsche':10},
            {'time':2, 'yugo':20, 'porsche':30},
            {'time':3, 'yugo':30, 'porsche':90}]
data = pd.DataFrame(raw_data)
# Compare Chart
porsche = hv.Dimension('porsche', label='Porsche', range=(0,100), unit='mph')
yugo = hv.Dimension('yugo', label='Yugo', range=(0,55), unit='mph')
time = hv.Dimension('time', label='Time', range=(0,3), unit='sec')
porsche_curve=hv.Curve(data, kdims=[time], vdims=[porsche], label=porsche.label)
porsche_dot=hv.Scatter(data, kdims=[time], vdims=[porsche], label=porsche.label)
yugo_curve=hv.Curve(data, kdims=[time], vdims=[yugo], label=yugo.label)
yugo_dot=hv.Scatter(data, kdims=[time], vdims=[yugo], label=yugo.label)( style={'marker':'s'})
cmp_chart = (porsche_curve * porsche_dot * yugo_curve * yugo_dot).relabel('Compare Auto Accelleration')(plot={'legend_position':'top_left'})
cmp_chart
```
<img width="241" alt="holoviewaxislabelexample" src="https://user-images.githubusercontent.com/18220857/33356035-3a1ee788-d488-11e7-99e5-925a3772f86f.png">

**Comment by jbednar** [comment]
I agree that the current behavior is misleading, and I already filed an issue to that effect (#1751). I'll take this issue as support for what I proposed there already!

However, in your case isn't the immediate problem that neither "Porsche" nor "Yugo" are actually valid descriptions of  dimensions?  Here the dimension for the y axis of all these elements should be something like ``speed = hv.Dimension('speed', label='Speed', range=(0,None), unit='mph')``; all of them plot speed in mph against time.  (Here None indicates that there is no upper bound on speed; there shouldn't usually be a need to supply an explicit bound like that in cases where the values aren't actually bounded (both for speed and time here, which are both only bounded on the lower side (at zero).)  Anyway, once you declare appropriate dimensions, this problem should go away.

**Comment by johnzzzz** [comment]
Looks like kdim and vdim parameters have two purposes that sometimes are at odds. 
First they define which column_id to lookup data in a table. 
Second they are used to specify the axis Dimension.  

Code using jbednar's recommendation to make generic Dimensions for the physical phenomena, with units, might look like this:
```python
# Compare Chart Explicit Data and Generalized Dim
speed_dim = hv.Dimension('speed', label='Velocity', range=(0,None), unit='mph')
time_dim = hv.Dimension('time', label='Time', range=(0,None), soft_range=(0,4), unit='sec')
porsche_curve=hv.Curve((data['time'], data['porsche']), kdims=[time_dim], vdims=[speed_dim], label='Porcshe')
yugo_curve=hv.Curve((data['time'], data['yugo']), kdims=[time_dim], vdims=[speed_dim], label='Yugo')
cmp_chart2 = (porsche_curve * yugo_curve).relabel('Compare Auto Accelleration')(plot={'legend_position':'top_left'})
cmp_chart2
```
![holoviewgeneraldimexample](https://user-images.githubusercontent.com/18220857/33394183-897eebc0-d50f-11e7-9787-f85ad3bb1e22.png)
Where as, using the Dimensions to access the data columns and axis labels could look like this:
```python
# Compare Chart Inferred Data and Inferred Dim
porsche_curve=hv.Curve(data, kdims=['time'], vdims=['porsche'], label='Porcshe')
yugo_curve=hv.Curve(data, 'time', 'yugo', label='Yugo')
cmp_chart3 = (porsche_curve * yugo_curve).relabel('Compare Auto Accelleration')(plot={'legend_position':'top_left'})
cmp_chart3
```
![holoviewinferreddimexample](https://user-images.githubusercontent.com/18220857/33394353-0e2ecd72-d510-11e7-9a8e-69c961f385f6.png)

Neither of these cases, allow me to explicity get at the charts yaxis and change the label to "mph". All the ways I have tried are indirect. Even trying to redim the speed_dim 
```python
cmp_chart2.redim.label(speed='Mps')
```
results in the error  "ValueError: Cannot override an existing Dimension label"

**Comment by johnzzzz** [comment]
The pragmatic question is how do I get directly at a charts y-axis's dimension?
So that I can change its label, and style

**Comment by philippjfr** [comment]
Before going into some more pragmatic answers I'll quickly comment on why it is the way it is at the moment. HoloViews is best suited towards visualizing tidy datasets where each column corresponds to a variable/dimension and can therefore easily be grouped. So rather than organizing data like this:

```
Time |   Yugo   | Porsche
-----|----------|---------
0    | 0.000000 | 0.000000
1    | 1.000000 | 1.000000
2    | 2.297397 | 3.482202
3    | 3.737193 | 7.224674
```

We generally recommend organizing your data like this:

```
Time |  Make   | Speed
-----|---------|---------
0    | Yugo    | 0.000000
1    | Yugo    | 1.000000
2    | Yugo    | 2.297397
...  |         |
0    | Porsche | 0.000000
1    | Porsche | 1.000000
2    | Porsche | 3.482202
```

Once your data is organized in this way there is no more tension between column names and axis labels. You can tidy your data using the ``pd.melt`` function and we will be providing means of easily converting datasets into a tidy format (see for example my suggestion in https://github.com/ioam/holoviews/pull/2164).

That said if you really want to take full control over axis labels you can always modify the plot objects directly by using hooks which modify the matplotlib/bokeh objects directly, here is a simple example setting the axis labels directly:

```python
def hook(plot, element):
    plot.handles['axis'].set_xlabel('A')
    plot.handles['axis'].set_ylabel('B')
    
hv.Curve([]).opts(plot=dict(final_hooks=[hook]))
```

**Comment by jbednar** [comment]
Good point; @philippjfr, I think you figured out what the real problem is.  Seems like we might need to spend more time talking about non-tidy datasets in the docs, so that it's clear what's going to be difficult when your data isn't tidy, and why making it tidy makes your life easier.  So I'll rename this into an issue about clarifying tidy vs. wide columnar datasets in the docs.

**Comment by johnzzzz** [comment]
Thank you for pointing out that holoviews assumes data is always "tidy".
I agree that conversion between tidy data sets and wide datasets is a conceptual hurdle;
which is at least as import as the separation of model (data) and view (visualization);
and plot (chart type) vs style (rendering).
I thought I could annotate the data columns, map the data columns to the views, annotate the views with a style, and holoviews would render it. The missing step is that the raw data must be normalized (tidied), before the columns are annotated.
The holoview annotation concept of "group", also seems to be part of Tidy-ness

One problem with the current documentation is it always shows the clever cases where the data is tidy and the  labels can be correctly inferred from the name. This helps motivate using holoviews for the novice, but hides the subtly of what is going on in the inference layer, which hinders the user's learning curve when the user needs to take control of the inference.  I would suggest that you can use the cleaver examples in the Getting Started section, but never use them in the main user guide

**Comment by jbednar** [comment]
> holoviews assumes data is always "tidy".

I agree that we should expand coverage of non-tidy datasets in the docs, but I think that's an overstatement. It's not so much that we *assume* data is tidy, it's that *if* your data is tidy, various things will be much easier to do. You can always plot even your non-tidy data if you explicitly pull out the chunks that you know relate to each other:

![image](https://user-images.githubusercontent.com/1695496/33696263-0939438a-dac7-11e7-93fc-466f27ed413d.png)

**Comment by jbednar** [comment]
Oops; I see you basically had this code already. Ok, here's a slightly expanded version showing how you can change the dimension if you want to do it after the fact:

![image](https://user-images.githubusercontent.com/1695496/33696557-6c0b2716-dac8-11e7-98ac-e2ebd9714347.png)

I don't think there's anything less well supported about this approach; it's just more verbose than with tidy data and not yet very well documented in the examples.
